base/broadcast.jl

```diff
@@ -691,8 +691,48 @@ eltypes(t::Tuple{Any}) = Tuple{_broadcast_getindex_eltype(t[1])}
 eltypes(t::Tuple{Any,Any}) = Tuple{_broadcast_getindex_eltype(t[1]), _broadcast_getindex_eltype(t[2])}
 eltypes(t::Tuple) = Tuple{_broadcast_getindex_eltype(t[1]), eltypes(tail(t)).types...}
 
+function promote_typejoin_union(::Type{T}) where T
+    if T === Union{}
+        return Union{}
+    elseif T isa Union
+        return promote_typejoin(promote_typejoin_union(T.a), promote_typejoin_union(T.b))
+    elseif T <: Tuple
+        return typejoin_union_tuple(T)
+    else
+        return T
+    end
+end
+
+@pure function typejoin_union_tuple(::Type{T}) where {T}
+    p = (Core.Compiler.unwrap_unionall(T)::DataType).parameters
```

Comment:
```suggestion
    p = (Base.unwrap_unionall(T)::DataType).parameters
```

Reply:
but might also be a `::Union` (for which both elements are Tuple), since `promote_typejoin_union` doesn't handle this case (of UnionAll wrappers)

Reply:
Sorry, not sure what I need to change here. :-)

Reply:
The type-assert—it's either a Union or DataType here

Reply:
Got it. I've pushed a fix.